VMQ refactoring & optimization

LKomaryanskiy opened the pull request:
This PR contains patches with fixes and optimizations of VMQ performance. Mainly, these changes introduce mechanisms to avoid package losses, reduce the number of Xen events to DomD, and as a result improve bandwidth.
Currently, the iperf tests are as follows:
- Default MTU (1500):
- - To host:
- - - (TCP)
iperf3 -c 192.168.1.100 -t 10 -b 0 -M 8134: 440 Mbits/sec
iperf3 -c 192.168.1.100 -t 10 -b 0 -M 8134 -R: 348 Mbits/sec
- - - (UDP)
iperf3 -c 192.168.1.100 -t 10 -u -b 0: 411 Mbits/sec
iperf3 -c 192.168.1.100 -t 10 -u -b 0 -R: 511 Mbits/sec
- - To VMQ:
- - - (TCP)
iperf3 -c 192.168.5.1 -t 10 -b 0 -M 8134: 744 Mbits/sec
iperf3 -c 192.168.5.1 -t 10 -b 0 -M 8134 -R: 907 Mbits/sec
- - - (UDP)
iperf3 -c 192.168.5.1 -t 10 -u -b 0: 425 Mbits/sec
iperf3 -c 192.168.5.1 -t 10 -u -b 0 -R: 881 Mbits/sec
- 9000 MTU:
- - To host:
- - - (TCP)
iperf3 -c 192.168.1.100 -t 10 -b 0 -M 8134: 731 Mbits/sec
iperf3 -c 192.168.1.100 -t 10 -b 0 -M 8134 -R: 700 Mbits/sec
- - - (UDP)
iperf3 -c 192.168.1.100 -t 10 -u -b 0 -l 8134: 637 Mbits/sec
iperf3 -c 192.168.1.100 -t 10 -u -b 0 -l 8134 -R: 783 Mbits/sec
- - To VMQ:
- - - (TCP)
iperf3 -c 192.168.5.1 -t 10 -b 0 -M 8134: 1.37 Gbits/sec
iperf3 -c 192.168.5.1 -t 10 -b 0 -M 8134 -R: 1.37 Gbits/sec
- - - (UDP)
iperf3 -c 192.168.5.1 -t 10 -u -b 0 -l 8134: 635 Mbits/sec
iperf3 -c 192.168.5.1 -t 10 -u -b 0 -l 8134 -R: 1.76 Gbits/sec


**With optimizations:**
- Default MTU (1500):
- - To host:
- - - (TCP)
iperf3 -c 192.168.1.100 -t 10 -b 0 -M 8134: 495 Mbits/sec (**+12.5%**)
iperf3 -c 192.168.1.100 -t 10 -b 0 -M 8134 -R: 583 Mbits/sec (**+67.5%**)
- - - (UDP)
iperf3 -c 192.168.1.100 -t 10 -u -b 0: 867 Mbits/sec (**+109.5%**)
iperf3 -c 192.168.1.100 -t 10 -u -b 0 -R: 891 Mbits/sec (**+74%**)
- - To VMQ:
- - - (TCP)
iperf3 -c 192.168.5.1 -t 10 -b 0 -M 8134: 894 Mbits/sec (**+20%**)
iperf3 -c 192.168.5.1 -t 10 -b 0 -M 8134 -R: 1.06 Gbits/sec (**+20%**)
- - - (UDP)
iperf3 -c 192.168.5.1 -t 10 -u -b 0: 943 Mbits/sec (**+122%**)
iperf3 -c 192.168.5.1 -t 10 -u -b 0 -R: 881 Mbits/sec (**+72.5%**)
- 9000 MTU:
- - To host:
- - - (TCP)
iperf3 -c 192.168.1.100 -t 10 -b 0 -M 8134: 764 Mbits/sec (**+4.5%**)
iperf3 -c 192.168.1.100 -t 10 -b 0 -M 8134 -R: 875 Mbits/sec (**+25%**)
- - - (UDP)
iperf3 -c 192.168.1.100 -t 10 -u -b 0 -l 8134: 958 Mbits/sec (**+50%**)
iperf3 -c 192.168.1.100 -t 10 -u -b 0 -l 8134 -R: 992 Mbits/sec (**+26.5%**)
- - To VMQ:
- - - (TCP)
iperf3 -c 192.168.5.1 -t 10 -b 0 -M 8134: 1.57 Gbits/sec (**+14.5%**)
iperf3 -c 192.168.5.1 -t 10 -b 0 -M 8134 -R: 1.52 Gbits/sec (**+11%**)
- - - (UDP)
iperf3 -c 192.168.5.1 -t 10 -u -b 0 -l 8134: 1.78 Gbits/sec (**+187%**)
iperf3 -c 192.168.5.1 -t 10 -u -b 0 -l 8134 -R: 1.78 Gbits/sec (**+- the same**)

**Checkpatch output**:
git diff HEAD~10 --patch | scripts/checkpatch.pl
total: 0 errors, 0 warnings, 0 checks, 638 lines checked

Your patch has no obvious style problems and is ready for submission.

LKomaryanskiy committed:
net: renesas: rswitch: add interdomain sync for VMQ tx->rx chains

Currently, DomU VMQ TX doesn't know the exact free space for descriptors
in DomD VMQ RX chain. As a result, it leads to RX overflowing and packet
loss during performance tests. Implementing synchronization via interdomain
shared memory with appropriate counters fixes this issue.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: fix uninitialized list

Currently, in case of disabling L3 offload while using VMQs in DomD
the system will hang, due to accessing to uninitialized list.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: reduce the number of TX events from DomU

Currently, every xmit_start invocation from DomU sends Xen event to
DomD to signalize DomD trigger TX chain and start DMA. This enormous
number of such events lead to slowing down the whole system including
rswitch_poll function. In turn, rswitch_poll is not able to process
all incoming packets, even if rescheduled all the time. So, to avoid
such scenario, implemented another way that reduces the number of Xen
events:

	1. Backend driver initializes high-resolution timer with
	   callback that periodically trigger TX chain;
	2. Frontend driver checks via shared memory if hr-timer
	   has not scheduled and if not send Xen event;
	3. Backend, after getting TX event, set flag, that timer
	   is scheduled and schedules timer in soft-IRQ context
	   with triggering chain;
	4. Backend periodically reschedules timer until DomU TXed
	   counter is bigger than DomD RXed counter. After all work
	   is done, sets back flag that DomD timer is ready to a new
	   schedule and stops hr-timer.

This mechanism reduces the total number of Xen events, as well as
hard-IRQs from Xen and reduces time of rswitch_poll execution that in
turn helps to process more packets at the same period of time. As a
result it increases bandwidth between DomU and DomD. Also, due to adding
new fields to vmq_info structure, the protocol version is bumped.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: add proper error handling in front driver

In case of any error, the allocated resources such as event channels,
netdevice, binded IRQ handlers should be properly freed or unbinded.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: unbind irq handlers in case of connection failure

In case of any error while connection, irq handlers should be unbinded.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: do not try to add fwd offload for VMQ devices

Currently, if traffic incomes from TSN to VMQ device, offloading
mechanism runs workqueue for every packet that slows down bandwidth
while forwarding from TSN to VMQ. This check can be done much earlier,
do not run workqueue and as a result, improve bandwidth.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: set frontend VMQ TX delay to zero

Delays were introduced to reduce the number of Xen events that
improves performance. Another mechanism has been implemented to
reduce Xen events to trigger TX chain, so delays for this chain
are not needed anymore.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: change default value for IRQ RX chain delay

After optimization, the delay value can be reduced from 0x100 to 0x40
and it currently gives better performance.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: send only appropriate Xen event in case of back IRQ

As we know from which chain TX or RX comes IRQ, we can trigger only
the appropriate event channel. It reduces the total amount of events
and improves performance.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: move memory barrier after all operations with DMA

The memory barrier should be set after all descriptors are ready. Only after
this, the trigger chain can be called. Currently, dma_wmb is placed before
setting the descriptors' die_dt field. In this case, when using multi-descriptors,
a situation may occur where the CPU reorders operations and triggers the chain
before all data writings are completed. This can lead to TX ring corruption,
where the FSTART descriptor was sent but FMID and FEND descriptors remain in
the ring buffer. Consequently, because the FSTART descriptor is missed, DMA
ignores the rest of the packet and does not write back that they were sent,
so we can not reuse them. As a result, the TX chain gets stuck because the
current index is 1024 greater than the dirty index.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

firscity reviewed (approved):
Amazing job, thank you!

lorc commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -2402,6 +2420,22 @@ static int rswitch_start_xmit(struct sk_buff *skb, struct net_device *ndev)
+					READ_ONCE(rdev->vmq_info->front_rx));
+		}
+		if (num_desc > rx_chain_free) {
+			ret = NETDEV_TX_BUSY;
```
Should we stop the queue in this case? I believe that `NETDEV_TX_BUSY` is the hard error and we really need to avoid signaling it.

LKomaryanskiy replied:
Good point! Added netif_stop_subqueue in the latest PR version.

lorc replied:
AFAIK, we need another check that the subqueue when we use last free descriptor. Again, returning `NETDEV_TX_BUSY` is the last resort.

Check other network drivers,  they are stopping subqueue earlier.

lorc commented on drivers/net/ethernet/renesas/rswitch.h:
```diff
@@ -306,6 +313,8 @@ struct rswitch_device {
 	struct device *vlan_parent;
 	bool mondev;
+	struct rswitch_vmq_status *vmq_info;
+	bool is_vmq;
```
I believe it is a good time to introduce some `enum rdev_type`. Because we already have `is_vmq`and `mondev` fields, implicit checks for being vlan or  front device... This is getting out of hand.

lorc commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -1329,7 +1343,6 @@ static bool rswitch_rx_chain(struct net_device *ndev, int *quota, struct rswitch
 		napi_gro_receive(&rdev->napi, skb);
 		rdev->ndev->stats.rx_packets++;
 		rdev->ndev->stats.rx_bytes += pkt_len;
-
```
Spurious change?

LKomaryanskiy replied:
Removed unnecessary newline removing

lorc commented on drivers/net/ethernet/renesas/rswitch.h:
```diff
@@ -274,6 +274,13 @@ struct rswitch_filters {
 };
 
+struct rswitch_vmq_status {
+	u64 front_tx, front_rx;
```
It is good idea to introduce some `protocol_version` field, so we at least can refuse to work if backend and frontend use different protocols  versions.

LKomaryanskiy replied:
Added one more patch, that introduces this kind of enum.

lorc reviewed (commented):
s/packages/packets in the commit messages

Also, I am wondering if it is possible to minimize use of high resolution timers. I have a gut feeling that we can leverage clever use of TX interrupts and TX interrupt delays, but it requires time to work it out and make some experiments. If you have spare time, we can discuss this.

GrygiriiS commented on drivers/net/ethernet/renesas/rswitch_xenback.c:
```diff
@@ -58,7 +58,7 @@ static int tx_trigger_time_nsec = 200 * 1000;
 MODULE_PARM_DESC(chain_irq_delay,
-		 "Set IRQ status delay for VMQ device RX/TX chains in range 0-0xfff");
+		 "Set IRQ status delay for VMQ device RX chain in range 0-0xfff");
 
```
It looks like smth. compatible with ` ethtool -C|--coalesce`?

LKomaryanskiy replied:
I have not used this parameter before, but according to ethtool documentation, it seems like it acts the same as R-Switch HW delays.

GrygiriiS commented on drivers/net/ethernet/renesas/rswitch.c:
```diff
@@ -1144,10 +1144,24 @@ void rswitch_enadis_rdev_irqs(struct rswitch_device *rdev, bool enable)
+		} else {
+			rswitch_vmq_front_trigger_tx(rdev);
+		}
+	}
```
It's probably not the same, but check `gro_flush_timeout` and 'napi_defer_hard_irqs'. FYI

LKomaryanskiy replied:
Thank you for this info, I read about these fields in the net device structure. This is different scenario because these lines aimed not for repolling napi, but for retriggering HW from DomD side.

GrygiriiS commented:
You posted 10 patches. And for some test cases it's declared up to +187% improvements. Are all patches affect on final result or only some of them? Which patches affects the most? In general, for each patch should be provided test result so they can be defined as high priority, while patches not affecting performance better to push as separate series "bug-fixes/improvements".

LKomaryanskiy committed:
net: renesas: rswitch: introduce enum with rdev types

There are plenty of rdev types such as TSN, monitoring, VMQ, VLAN
devices. Currently, in some cases or for the further changes we have
to identify exact type of device to act accordinally. Introducing
enum with rdev type will reduce amount of bool variables.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

LKomaryanskiy commented:
@firscity @lorc @GrygiriiS First of all, thank you for your review comments.
I apologize for such delay, I had to spend time on some research related to using HW delays instead of HR timers. Unfortunately, it doesn't work as expected and due to HW-specific behavior, we cannot properly use HW IRQ delays instead of HR-timers.

Please, check new version of PR. I made the following changes compared with previous version:

> s/packages/packets in the commit messages

1. Fixed commit messages: changed "packages" to "packets";

> Should we stop the queue in this case? I believe that NETDEV_TX_BUSY...

2. Added netif_stop_subqueue in case if domain RX chain doesn't have enough space for new descriptors. Previously I had issues with this, but current version works fine.

> I believe it is a good time to introduce some enum rdev_type..

3. Added one more patch, that introduces this kind of enum.

> Spurious change?

4. Removed unnecessary newline removing;

> It is good idea to introduce some protocol_version field..

5. Added one more field to VMQ info structure to make sure that frontend and backend use the same version of this structure;
6. Reordered some patches: bugfixes and "preparation" patches for the main changes placed at the beginning of PR. On the other hand, the patches with improving performance are placed at the end of PR.

Also, please find below the answers to questions:

> It looks like smth. compatible with ethtool -C|--coalesce?

I have not used this parameter before, but according to ethtool documentation, it seems like it acts the same as R-Switch HW delays. 

> It's probably not the same, but check gro_flush_timeout and 'napi_defer_hard_irqs'

Thank you for this info, I read about these fields in the net device structure. This is different scenario because these lines aimed not for repolling napi, but for retriggering HW from DomD side.

> Are all patches affect on final result or only some of them?

I reordered some patches to make PR more clear and easier to review. Now, bugfixes and "preparation" patches for the main changes are placed at the beginning of PR. On the other hand, the patches with improving performance placed at the end of PR started from:
https://github.com/xen-troops/linux/pull/164/commits/9a0464eeb0736bceca6b1977b709a3a6dfa33c31

> Which patches affects the most?

The following patch affects the most:
https://github.com/xen-troops/linux/pull/164/commits/9a0464eeb0736bceca6b1977b709a3a6dfa33c31
But without next patches, that remove or reduce IRQ delays and as a result, remove limits bandwith we can not test this patch on his full power:) On the other hand, we also can not place changing delays before this patch, because it will lead to bigger load and as a result, if somebody will checkout at the patch in the middle of the tree he gets broken build, which will have worse performance.

> provided test result so they can be defined as high priority

It's also not so simple to do, because as I wrote in the previous answer, we can not easily change the current sequence of patch seria. Also, if we measure the performance diff between patches, we got wrong conclusion that 
https://github.com/xen-troops/linux/pull/164/commits/143a97d8b36dfbeb46eab717363d5b12f1fefe8a
improves performance the best. But it's wrong because all that does this patch is removing limits for bandwidth. 

> push as separate series "bug-fixes/improvements"

I agree. Next time I will try to separate PR for smaller series.

Also, here is additional information requested in personal messages about the number of IRQs and packet loss:
1. With the current implementation, we have 1 hard-irq in DomD for every TX. With this PR, the number of hard IRQs reduced in ~26-27 times:
Transmitting between domains 856459 datagrams for 10 seconds generates only 32242 xen events (hard IRQs). Also, I have to admit, that with this PR additionally, we have 32146 HR timer soft IRQs, that were not before. Nevertheless, it's definitely better, than we have now.
2. Previously, we had packet loss of 15-40% depending on the testcase. This PR introduces mechanism to prevent overflowing chains and as a result packet loss is close to zero.

firscity commented on drivers/net/ethernet/renesas/rswitch.h:
```diff
@@ -216,7 +216,7 @@ struct rswitch_gwca_chain {
 #define MAX_MTU_SZ		9000
 #define MAX_DESC_SZ		2048
-#define VMQ_INFO_VERSION	1
+#define VMQ_INFO_VERSION	2
```
nit: I think it is worth to mention this in commit message ("due to changes in ... needed to bump ...")

LKomaryanskiy replied:
Thank you for the review. It's a good point - I will add this to commit message.

firscity reviewed (approved):
with one nit:

`Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>`

LKomaryanskiy commented:
In the new version of PR:
Added RB tag from @firscity and added info to commit message about changing version of vmq_info protocol.

lorc reviewed (commented):
One more question: did you validated use-case with the passed-thru TSNx to DomU? Is it still working?

LKomaryanskiy committed:
net: renesas: rswitch: stop subqueue if it is the last free descriptor

At the end of transmission we can check if we have enough free tx
descriptors for the next transmition and if we don't - we can
stop subqueue without returning NETDEV_TX_BUSY error.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

LKomaryanskiy committed:
net: renesas: rswitch: initialize be->rdev pointer in case of TSN

Currently, this pointer is uninitialized. In case someone needs
to use it, the null pointer dereference error occurs.

Signed-off-by: Leonid Komarianskyi <leonid_komarianskyi@epam.com>
Reviewed-by: Dmytro Firsov <dmytro_firsov@epam.com>
Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

LKomaryanskiy commented:
@lorc  thank you for review.
I added one more patch with stopping subqueue on the same xmit call, when we used the last descriptor. Also, I made appropriate changes in https://github.com/xen-troops/linux/pull/164/commits/b7e4003ab511cc5bcb8d7e2f8d9d70624a151bf3.

Regarding TSN usage: I rechecked workability with this PR and found an issue during initialization, that wasn't visible before. In case of using TSNs in DomU, `be->rdev` pointer wasn't initialized, but in this PR I added code that uses it. To fix this issue, I added patch with initialization of `be->rdev`. Now, all works fine.

@firscity @lorc  please, take a look at the new version of PR.

lorc reviewed (approved):
So, my intuition was right :)

Thank you

`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`

LKomaryanskiy commented:
@lorc thank you and your intuition :D 
It surprised me, because the first version of PR works fine with TSN, even with uninitialized rdev, but with different checks. Nevertheless, thank you again. 
Added @lorc 's RB tag.
